File: src/main/java/id/walt/ssikitexamples/DidEbsi.java
package id.walt.ssikitexamples;

import id.walt.crypto.KeyAlgorithm;
import id.walt.model.DidMethod;
import id.walt.servicematrix.ServiceMatrix;
import id.walt.services.did.DidService;
import id.walt.services.ecosystems.essif.EssifClient;
import id.walt.services.ecosystems.essif.didebsi.DidEbsiService;
import id.walt.services.key.KeyService;

public class DidEbsi {

    private final KeyService keyService = KeyService.Companion.getService();
    private final DidEbsiService didEbsiService = DidEbsiService.Companion.getService();

    ///////////////////////////////////////////////////////////////////////////
    // PREREQUISITE
    // Place token from https://app.preprod.ebsi.eu/users-onboarding/ in file data/ebsi/bearer-token.txt
    ///////////////////////////////////////////////////////////////////////////

    public static void main(String[] args) {
        new DidEbsi().run();
    }

    public void run() {

        // Load walt.id SSI Kit services from "$workingDirectory/service-matrix.properties"
        new ServiceMatrix("service-matrix.properties");

        System.out.println("SSI Kit example how to create a DID EBSI");

        ///////////////////////////////////////////////////////////////////////////
        // Create an asymmetric keypair
        ///////////////////////////////////////////////////////////////////////////

        var keyId = keyService.generate(KeyAlgorithm.EdDSA_Ed25519);
        System.out.println("EBSI key generated: " + keyId);

        var ethKeyId = keyService.generate(KeyAlgorithm.ECDSA_Secp256k1);
        System.out.println("ETH key generated: " + ethKeyId);

        ///////////////////////////////////////////////////////////////////////////
        // Create an EBSI compliant Decentralized Identifier
        ///////////////////////////////////////////////////////////////////////////

        var didEbsi = DidService.INSTANCE.create(DidMethod.ebsi, keyId.getId(), null); // Note, that the DID is stored in /data/did/create

        System.out.println("EBSI DID created : " + didEbsi);

        var didDoc = DidService.INSTANCE.loadDidEbsi(didEbsi);
        System.out.print("\nDID EBSI Document loaded:\n" + didDoc.encodePretty());


        EssifClient.INSTANCE.onboard(didEbsi, null);

        EssifClient.INSTANCE.authApi(didEbsi);

        didEbsiService.registerDid(didEbsi, ethKeyId.getId());

        System.out.print("\nDID EBSI Document successfully registered:\n" + didEbsi);

        System.out.print("\nCheck EBSI API at https://api.preprod.ebsi.eu/docs/?urls.primaryName=DID%20Registry%20API:\n");


    }
}

File: src/main/kotlin/id/walt/ssikitexamples/DidEbsi.kt
package id.walt.ssikitexamples

import id.walt.crypto.KeyAlgorithm
import id.walt.model.DidMethod
import id.walt.servicematrix.ServiceMatrix
import id.walt.services.did.DidKeyCreateOptions
import id.walt.services.did.DidService
import id.walt.services.ecosystems.essif.EssifClient
import id.walt.services.ecosystems.essif.didebsi.DidEbsiService
import id.walt.services.key.KeyService
import mu.KotlinLogging

fun main(){
    ebsiWorkflow()
}

fun ebsiWorkflow(){
    // Load SSI-Kit services
    val log = KotlinLogging.logger {}
    ServiceMatrix("service-matrix.properties")

    val keyService = KeyService.getService()
    val didEbsiService = DidEbsiService.getService()

    ///////////////////////////////////////////////////////////////////////////
    // PREREQUISITE
    // Place token from https://app-pilot.ebsi.eu/users-onboarding/v2 in file data/ebsi/bearer-token.txt
    ///////////////////////////////////////////////////////////////////////////

    // 1. Create key
    val keyId = keyService.generate(KeyAlgorithm.ECDSA_Secp256k1)
    val ethKeyId = keyService.generate( KeyAlgorithm.ECDSA_Secp256k1 )

    // 2. Create did
    val didEbsi = DidService.create( DidMethod.ebsi, keyId.id, DidKeyCreateOptions(useJwkJcsPub = true) )
    log.debug{ "${didEbsi}"}

    val didDoc = DidService.loadDidEbsi( didEbsi )
    log.debug { didDoc.encodePretty() }

    // 3. Onboard
    val onboardDid = EssifClient.onboard( didEbsi, null )
    log.debug{ onboardDid }

    // 4. Authorize
    val authorizeDid = EssifClient.authApi( didEbsi )
    log.debug{ authorizeDid }

    // 5. Register
    val registerDid = didEbsiService.registerDid( didEbsi, ethKeyId.id )
    log.debug{ registerDid }

    // 6. Resolve
    val resolveDid = DidService.resolve( didEbsi )
    log.debug{ resolveDid }

}